llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -18200,6 +18200,19 @@ static SDValue performConcatVectorsCombine(SDNode *N,
     return true;
   };
 
+  // concat(uhadd(a, b), uhadd(c, d)) -> uhadd(concat(a, c), concat(b, d))
```

Comment:
There is some code that also looks at uhadd but with extract_subvector. Could that be made more generic to perform this fold more generally? We could expand it in the future to handle other operations too.

Do you mind also adding a check that the dest size is 128bits, to make sure this doesn't trigger for SVE fixed length vector too.

Reply:
> Do you mind also adding a check that the dest size is 128bits, to make sure this doesn't trigger for SVE fixed length vector too.
Yeah, that makes sense, I'll add the check.

Reply:
Can this optimisation also work for dests that are 64 bits too? I'm thinking of cases like this:

concat(v4i8 uhadd(v4i8 a, v4i8 b), v4i8 uhadd(v4i8 c, v4i8 d))

where the result is actually a legal NEON vector - v8i8? I imagine doing the concats on the inputs first will help with codegen too because the uhadd operation will then be legal I think.

Reply:
> Can this optimisation also work for dests that are 64 bits too? I'm thinking of cases like this:
> 
> concat(v4i8 uhadd(v4i8 a, v4i8 b), v4i8 uhadd(v4i8 c, v4i8 d))
> 
> where the result is actually a legal NEON vector - v8i8? I imagine doing the concats on the inputs first will help with codegen too because the uhadd operation will then be legal I think.

I think for legal NEON vectors there isn't a concat used so the optimisation won't apply for 64-bits.